hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocation.java

```diff
@@ -221,15 +241,44 @@ public void setInvocationTimeoutMillis(long invocationTimeoutMillis) {
         this.invocationTimeoutMillis = invocationTimeoutMillis;
     }
 
-
     /**
+     * We make sure that `notifyResponse` can not be called by multiple threads concurrently
+     * Only ones that can set a null on `sentConnection` can notify the invocation
+     *
      * @param clientMessage return true if invocation completed
      */
     void notify(ClientMessage clientMessage) {
         assert clientMessage != null : "response can't be null";
         int expectedBackups = clientMessage.getNumberOfBackupAcks();
 
-        notifyResponse(clientMessage, expectedBackups);
+        if (getPermissionToNotify(clientMessage.getCorrelationId())) {
+            assert sentConnection == null;
```

Comment:
Minor: It is just a micro-optimization but we can move `int expectedBackups = clientMessage.getNumberOfBackupAcks();`
to the if block so that we are not reading it when we don't need it